### PR TITLE
UI: Use Client Count export API

### PR DESCRIPTION
**Pull request by hashishaw**
### Description
This PR updates the Client Count Dashboard to use the dedicated [export API](https://developer.hashicorp.com/vault/api-docs/system/internal-counters#activity-export). Relies on enterprise PR 6171

- Copy for the export modal was updated to reflect the new source of data
- Added loading state to modal when export is running
- Added a dropdown to the modal for the user to choose the format
- The current namespace + selected namespace is added to the export header so that the downloaded data reflects the filters on the page
- Export button is hidden for users without `sudo` capabilities to export at the given namespace

Kapture 2024-07-10 at 14 49 58

**Updated modal content**
updated copy
loading

**Updated CSV content**
Screenshot 2024-06-12 at 12 26 00

**Previous CSV content**
Screenshot 2024-06-12 at 12 27 04

- [x] Ent tests pass

### CODE REVIEW

**Review by hellobontempo — COMMENTED** [review]
Just a comment about a conditional. Great work!

**Review by hellobontempo — APPROVED** [review]
Could you add a screenshot of how the new export csv data looks?

File: ui/app/utils/query-param-string.js [hunk 1]
```diff
@@ -0,0 +1,18 @@
+/**
+ * queryParamString converts an object to a query param string with URL encoded keys and values.
+ * It does not include values that are falsey.
+ * @param {object} queryObject with key-value pairs of desired URL params
+ * @returns string like ?key=val1&key2=val2
+ */
+export default function queryParamString(queryObject) {
+  if (typeof queryObject !== 'object') return '';
```
[the remark this answers is not in the document]
Comment (hashishaw) [reply ui/app/utils/query-param-string.js:8]: Good thinking!

File: ui/app/routes/vault/cluster/clients/counts.ts [hunk 2]
```diff
@@ -33,7 +31,7 @@ export default class ClientsCountsRoute extends Route {
   queryParams = {
     start_time: { refreshModel: true, replace: true },
     end_time: { refreshModel: true, replace: true },
-    ns: { refreshModel: false, replace: true },
+    ns: { refreshModel: true, replace: true },
```
Comment (hashishaw) [inline ui/app/routes/vault/cluster/clients/counts.ts:34]: This is necessary so the `getExportCapabilities` method runs when the namespace changes.

**Review by hellobontempo — COMMENTED** [review]
Looking great! Just a couple comments/questions for now! Will continue reviewing once tests updates are complete 😄

File: ui/app/components/clients/attribution.js [hunk 3]
```diff
@@ -130,99 +160,19 @@ export default class Attribution extends Component {
...
-            : null;
-          const mountTotalClients = this.destructureCountsToArray(mount);
-          const mountNewClients = newMountData ? this.destructureCountsToArray(newMountData) : null;
-          csvData.push(this.constructCsvRow(namespace, mount, mountTotalClients, mountNewClients));
-        });
-      }
+  async generateCsvData() {
+    const adapter = this.store.adapterFor('clients/activity');
+    const currentNs = this.namespace.path;
+    const { startTimestamp, endTimestamp, selectedNamespace } = this.args;
```
[the remark this answers is not in the document]
Comment (hashishaw) [reply ui/app/components/clients/attribution.js:166]: Correct!

File: ui/app/components/clients/attribution.js [hunk 4]
```diff
@@ -44,7 +47,33 @@ import { format, isSameMonth } from 'date-fns';
...
+  }
+
+  @waitFor
+  async getExportCapabilities(ns = '') {
+    try {
+      // selected namespace usually ends in /
+      const url = ns
+        ? `${sanitizePath(ns)}/sys/internal/counters/activity/export`
+        : 'sys/internal/counters/activity/export';
+      const cap = await this.store.findRecord('capabilities', url);
```
[the remark this answers is not in the document]
Comment (hashishaw) [reply ui/app/components/clients/attribution.js:70]: I tested with this policy and it worked -- I think it grants read, etc under the hood

File: ui/app/components/clients/attribution.js [hunk 5]
```diff
@@ -130,99 +160,19 @@ export default class Attribution extends Component {
     }
   }
 
-  destructureCountsToArray(object) {
-    // destructure the namespace object  {label: 'some-namespace', entity_clients: 171, non_entity_clients: 20, acme_clients: 6, secret_syncs: 10, clients: 207}
-    // to get integers for CSV file
-    const { clients, entity_clients, non_entity_clients, acme_clients, secret_syncs } = object;
-    const { isSecretsSyncActivated } = this.args;
```
[the remark this answers is not in the document]
Comment (hashishaw) [reply ui/app/components/clients/attribution.js:137]: Sure, didn't realize this was the only place it was used

File: ui/app/components/clients/attribution.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (hellobontempo) [reply ui/app/components/clients/attribution.js:69]: Okay cool! Thanks for confirming

File: ui/app/components/clients/attribution.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (hellobontempo) [reply ui/app/components/clients/attribution.js:137]: Definitely double check me! When I reviewed it looked like it was no longer used

File: ui/app/adapters/clients/activity.js [hunk 6]
```diff
@@ -46,6 +47,30 @@ export default class ActivityAdapter extends ApplicationAdapter {
...
+    })}`;
+    let errorMsg;
+    try {
+      const options = query?.namespace ? { namespace: query.namespace } : {};
+      const resp = await this.rawRequest(url, 'GET', options);
+      if (resp.status === 200) {
+        return resp.blob();
+      }
+      // If it's an empty response (eg 204), there's no data so return an error
+      errorMsg = 'no data to export in provided time range.';
```
Comment (Monkeychip) [inline ui/app/adapters/clients/activity.js:64]: NIT: caps 'no' -> 'No'

**Review by Monkeychip — APPROVED** [review]
🤩 nice work!

File: ui/app/adapters/clients/activity.js (hunk 6 above)
Comment (hashishaw) [reply ui/app/adapters/clients/activity.js:64]: I'm matching the casing of the API error messages, which don't have capitalization

**Pull request merged**